zebra-network/src/address_book.rs

```diff
@@ -104,82 +117,98 @@ impl AddressBook {
         peers
     }
 
     /// Returns true if the address book has an entry for `addr`.
-    pub fn contains_addr(&self, addr: &SocketAddr) -> bool {
+    pub fn contains_addr(&self, addr: SocketAddr) -> bool {
         let _guard = self.span.enter();
-        self.by_addr.contains_key(addr)
+        self.by_addr.contains_key(&addr)
     }
 
-    /// Returns the entry corresponding to `addr`, or `None` if it does not exist.
+    /// Returns the entry corresponding to `addr`, or [`None`] if it does not exist.
     pub fn get_by_addr(&self, addr: SocketAddr) -> Option<MetaAddr> {
         let _guard = self.span.enter();
         self.by_addr.get(&addr).cloned()
     }
 
-    /// Add `new` to the address book, updating the previous entry if `new` is
-    /// more recent or discarding `new` if it is stale.
+    /// Apply `change` to the address book.
+    ///
+    /// If an entry was added or updated, return it.
     ///
     /// # Correctness
     ///
-    /// All new addresses should go through `update`, so that the address book
-    /// only contains valid outbound addresses.
-    pub fn update(&mut self, new: MetaAddr) {
+    /// All address book changes should go through `update`, so that the
+    /// address book only contains valid outbound addresses.
+    pub fn update(&mut self, change: MetaAddrChange) -> Option<MetaAddr> {
         let _guard = self.span.enter();
         trace!(
-            ?new,
+            ?change,
             total_peers = self.by_addr.len(),
             recent_peers = self.recently_live_peers().count(),
         );
+        let addr = change.get_addr();
+        let book_entry = self.by_addr.get(&addr).cloned();
 
         // If a node that we are directly connected to has changed to a client,
         // remove it from the address book.
-        if new.is_direct_client() && self.contains_addr(&new.addr) {
+        if change.is_direct_client(book_entry) && self.contains_addr(addr) {
             std::mem::drop(_guard);
-            self.take(new.addr);
-            return;
+            self.take(addr);
+            self.update_metrics();
+            return None;
         }
 
         // Never add unspecified addresses or client services.
         //
         // Communication with these addresses can be monitored via Zebra's
         // metrics. (The address book is for valid peer addresses.)
-        if !new.is_valid_for_outbound() {
-            return;
+        if !change.is_valid_for_outbound(book_entry) {
+            return None;
         }
 
-        if let Some(prev) = self.get_by_addr(new.addr) {
-            if prev.get_last_seen() > new.get_last_seen() {
-                return;
-            }
+        let new_entry = change.into_meta_addr(book_entry);
+        if let Some(new_entry) = new_entry {
+            self.by_addr.insert(addr, new_entry);
+            std::mem::drop(_guard);
+            self.update_metrics();
+            Some(new_entry)
+        } else {
+            None
         }
```

Comment:
This check skips invalid updates, see `MetaAddrChange::into_meta_addr` for details.

Currently, the only invalid updates are:
- attempted, success, or failed -> never attempted....
- never attempted gossiped or never attempted alternate -> never attempted seed

The "more recent last seen" check was wrong.

It is fixed by `into_meta_addr`, which:
* replaces untrusted times and services with direct or local data
* otherwise, keeps the first times and services we got, so that malicious peers can't keep updating them

Reply:
This change is unlikely to be causing test failures, because:
* all seed peers get a connection attempt